cassiopeia/simulator/SpatialSampler.py

```diff
@@ -0,0 +1,315 @@
+                leaves must have the `attribute_key` attribute that contains
+                2D spatial coordinates.
+            attribute_key: Node attribute that contains the spatial coordinates.
+                Defaults to `spatial`.
+            collapse_source: The source node from which to collapse unifurcations
+            collapse_duplicates: Whether or not to collapse duplicated character
+                states, so that only unique character states are present in each
+                ambiguous state. Defaults to True.
+
+        Raises:
+            CassiopeiaTreeError if any of the leaves does not have the
+                `attribute_key` attribute
+            LeafSubsamplerError if not all spatial coordinates are 2-dimensional
+                or if any of the leaves have ambiguous states
+        """
+        locations = {}
+        min_x, max_x = np.inf, -np.inf
+        min_y, max_y = np.inf, -np.inf
+
+        for leaf in tree.leaves:
+            if tree.is_ambiguous(leaf):
+                raise LeafSubsamplerError(f"Leaf {leaf} is ambiguous.")
+            location = tree.get_attribute(leaf, attribute_key)
+            if len(location) != 2:
+                raise LeafSubsamplerError(
+                    f"All coordinates must be 2-dimensional, but leaf {leaf} "
+                    f"has {len(location)}-dimensional coordinates."
+                )
+            locations[leaf] = location
+            min_x = min(min_x, location[0])
+            max_x = max(max_x, location[0])
+            min_y = min(min_y, location[1])
+            max_y = max(max_y, location[1])
+
+        # Construct grid and cells
+        spots = self.create_spots(
+            (min_x, max_x + self.cell_size),
+            (min_y, max_y + self.cell_size),
+        )
+        cells = {
+            leaf: self.create_cell(location[0], location[1])
+            for leaf, location in locations.items()
+        }
+
+        # Capture cells in each spot.
+        # This dictionary will contain the index of each spot (in the spots
+        # list) as the keys and a dictionary of captured cells (as returned by
+        # `capture_cells_on_spot`) as values.
+        captured_spots = {}
+        for i, spot in enumerate(spots):
+            captured = self.capture_cells(spot, cells)
+            if captured:
+                captured_spots[i] = captured
+
+        # Construct a new network and character matrix
+        sampled_tree = tree.copy()
+        spot_leaves = {}
+        for spot_i, captured in captured_spots.items():
+            spot_uuid = str(uuid.uuid4())
+            spot_leaves[spot_uuid] = spot_i
+            spot_states = self.sample_states(tree, captured)
+
+            # Add new leaf
+            captured_leaves = list(captured.keys())
+            new_time = sum(
+                tree.get_time(leaf) for leaf in captured_leaves
+            ) / len(captured_leaves)
```

Comment:
I think our trees are assumed to be ultrametric, so the leaves should all share the same time?